src/Workspaces/Core/MSBuild/MSBuild/DiagnosticReporter.cs

```diff
@@ -38,6 +41,11 @@ public void Report(DiagnosticReportingMode mode, string message, Func<string, Ex
             }
         }
 
+        internal void AddDiagnostic(WorkspaceDiagnostic diagnostic)
+        {
+            ImmutableInterlocked.Update(ref Diagnostics, d => d.Add(diagnostic));
```

Comment:
💡 If allocations are important here, you can avoid the capture using the overload of `Update` that takes a `TArg` state object.